proposed/logging-generator.md

```diff
@@ -0,0 +1,327 @@
+        return true;
+    }
+
+    public KeyValuePair<string, object?> this[int index]
+    {
+        get => ... // depends on the type used to store key value pairs
+    }
+
+    public IEnumerator<KeyValuePair<string, object>> GetEnumerator()
+    {
+        for (int i = 0; i < Count; ++i)
+        {
+            yield return this[i];
+        }
+    }
+
+    IEnumerator IEnumerable.GetEnumerator()
+    {
+        return GetEnumerator();
+    }
+}
+```
+
+In order to make use of a builder like `CustomLoggerParamsBuilder`, we would need new overloads in the future for LogInformation (and other similar APIs):
+
+```c#
+// We need new overloads to support this call:
+logger.LogInformation($"Welcome to {city}!");
+```
+
+But the sample below shows an application of the builder using an existing logging API:
+```c#
+var city = "Vancouver";
+
+logger.Log(
+    LogLevel.Information,
+    new EventId(1),
+    $"Welcome to {city}!",
+    new Exception(),
+    (s, e) => s);
+```
+The compiler would translate the above call to:
+```c#
+logger.Log<CustomLoggerParamsBuilder>(
+    LogLevel.Information,
+    new EventId(1),
+    builder, // the CustomLoggerParamsBuilder
+    new Exception(),
+    builder.FormatCallback
+);
+```
+Through this new compiler feature, the builder would be able to recognize name holes, (here the city) or even format specifiers. However, currently the proposed builder approach is limited in its ability to get both name holes and format specifiers at once. The logger to is able to identify both of these from message templates and would be good to have this feature available.
+
+Also, an older compiler, the builder overload may not be detectable and therefore the above log APIs could end up calling the wrong overload, which would cause in logging messages to completely lose the structure provided in the input. This could be however be mitigated with an analyzer that warns user when they use string interpolated messages on older compiler versions.
+ 
+### Limitations with string interpolation builder approach:
+
+- With the interpolated string builder approach, there is currently no way to get both name holes and format specifiers, but this is supported today already by the existing logger APIs.
+
+- The biggest concern for using the custom struct builder for our logging approach is that we'd need to store some value for name holes eagerly in the builder. This step cannot be done lazily because the structure we generate could ultimately get serialized in different ways decided by the different consumers of `ILogger.Log` call. The end result of using the builder this way, could be less efficient than what we already have in our logging APIs, even though our current approaches already do boxing.
+
+### Benefits:
+
+- The name hole approach adds enough value that would be good to consider in the future for cases where we are given an interpolated string rather than message templates.
```

Comment:
There are key benefits not called out here.  In particular, the interpolated string approach means callers don't need to guard the call sites if any computation is necessary to produce the arguments to the logging.  With all of the other approaches considered in this design, any such work done at the call site to produce the arguments will need to be guarded behind appropriate logging enabled checks, or else it'll end up being done even if logging is disabled.

Reply:
With this issue https://github.com/dotnet/runtime/issues/45290, there were new Define overloads added that guard the call site producing the arguments with isEnabled check

Reply:
@maryamariyan can you write out a sample of the latest code gen for the source generator.

Reply:
An example would be very helpful, as I'm currently missing how those APIs help here.  Just to take a concrete example, let's pick this method from ASP.NET:
https://github.com/dotnet/aspnetcore/blob/277cc564ce5a76c6613f29dd92c0168e06b41b1c/src/Shared/CertificateGeneration/CertificateManager.cs#L790-L791
Call site looks like:
https://github.com/dotnet/aspnetcore/blob/277cc564ce5a76c6613f29dd92c0168e06b41b1c/src/Shared/CertificateGeneration/CertificateManager.cs#L85-L88
Now obviously that's an event on an EventSource, but let's pretend it's not... with these logging APIs it would be translated to something like:
```C#
[LoggerMessage(2, LogLevel.Verbose, "Certificates: `{matchingCertificates}`")]
public static partial void DecribeFoundCertificates(this ILogger logger, string matchingCertificates);
```
And the call site would presumably be:
```C#
logger.DescribeFoundCertificates(ToCertificateDescription(matchingCertificates));
```
How does that caller avoid the `ToCertificateDescription(matchingCertificates)` computation without adding their own guard check?

Reply:
These APIs don't help there. You still need an if check here but I don't know why that's a problem...

Reply:
> You still need an if check here but I don't know why that's a problem...

Because people forget and then we ship .NET 5 with ASP.NET paying to log data about all certificates on the box even though that logging is disabled :wink:

Reply:
LOL, I think we should fix this code regardless.

Reply:
https://github.com/dotnet/aspnetcore/pull/27956

Reply:
I'm sure there's more

Reply:
> @maryamariyan can you write out a sample of the latest code gen for the source generator.

here's a sample:

```csharp
    static partial class Log
    {
        [LoggerMessage(32, LogLevel.Trace, "Certificates: `{matchingCertificates}`")]
        public static partial void DescribeFoundCertificates(this ILogger logger, string matchingCertificates);
    }
```

generates:
```csharp

namespace Example
{
    partial class Log
    {
        [global::System.CodeDom.Compiler.GeneratedCodeAttribute("Microsoft.Extensions.Logging.Generators", "1.0.0.0")]
        private static readonly global::System.Action<global::Microsoft.Extensions.Logging.ILogger, string, global::System.Exception?> _DescribeFoundCertificatesCallback =
            global::Microsoft.Extensions.Logging.LoggerMessage.Define<string>(global::Microsoft.Extensions.Logging.LogLevel.Trace, new global::Microsoft.Extensions.Logging.EventId(32, nameof(DescribeFoundCertificates)), "Certificates: `{matchingCertificates}`", true);

        [global::System.CodeDom.Compiler.GeneratedCodeAttribute("Microsoft.Extensions.Logging.Generators", "1.0.0.0")]
        public static partial void DescribeFoundCertificates(this Microsoft.Extensions.Logging.ILogger logger, string matchingCertificates)
        {
            if (logger.IsEnabled(global::Microsoft.Extensions.Logging.LogLevel.Trace))
            {
                _DescribeFoundCertificatesCallback(logger, matchingCertificates, null);
            }
        }

    }
}
```

usage:
```csharp
logger.DescribeFoundCertificates("some matching certificates go here!");
```

Reply:
> usage

My point is the usage with the source generator needs to be:
```C#
if (logger.IsEnabled(LogLevel.Trace))
{
    logger.DescribeFoundCertificates(Describe(matchingCertificates));
}
```

Reply:
For this specific usage yes. We don't need to delete the feature because one of these usages requires a manual if check.

Reply:
I'm not arguing we need to "delete the feature".  I'm arguing there's a real pit of failure here's that being swept under the rug.

Reply:
I don't think it's being swept under the rug. Typically the way we solve that pattern is by adding a method that takes unprocessed data and manually adds the if check. It's the exact same thing you need to do today, this doesn't improve that situation. I'm not sure how to fix that other than having a lambda. There are lots of lots that don't require complex transformation and are just logging the data that you have. Maybe an analyzer can detect these use cases and recommend an IsEnabled check.